coordinator/internal/authority/credentials.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2024 Edgeless Systems GmbH
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package authority
+
+import (
+	"context"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net"
+	"time"
+
+	"github.com/edgelesssys/contrast/internal/atls"
+	"github.com/edgelesssys/contrast/internal/attestation/snp"
+	"github.com/edgelesssys/contrast/internal/logger"
+	"github.com/edgelesssys/contrast/internal/memstore"
+	"github.com/google/go-sev-guest/proto/sevsnp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc/credentials"
+	"k8s.io/utils/clock"
+)
+
+// Credentials are gRPC transport credentials that dynamically update with the Coordinator state.
+type Credentials struct {
+	issuer   atls.Issuer
+	getState func() (*State, error)
+
+	logger                     *slog.Logger
+	attestationFailuresCounter prometheus.Counter
+	kdsGetter                  *snp.CachedHTTPSGetter
+}
+
+// Credentials creates new transport credentials that validate peers according to the latest manifest.
+func (a *Authority) Credentials(reg *prometheus.Registry, issuer atls.Issuer) (*Credentials, func()) {
+	ticker := clock.RealClock{}.NewTicker(24 * time.Hour)
+	kdsGetter := snp.NewCachedHTTPSGetter(memstore.New[string, []byte](), ticker, logger.NewNamed(a.logger, "kds-getter"))
+	attestationFailuresCounter := promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Subsystem: "contrast_meshapi",
+		Name:      "attestation_failures_total",
+		Help:      "Number of attestation failures from workloads to the Coordinator.",
+	})
+
+	return &Credentials{
+		issuer: issuer,
+		getState: func() (*State, error) {
+			if err := a.syncState(); err != nil {
+				return nil, fmt.Errorf("syncing state: %w", err)
+			}
+			state := a.state.Load()
+			if state == nil {
+				return nil, errors.New("coordinator is not initialized")
+			}
+			return state, nil
+		},
+		logger:                     a.logger,
+		attestationFailuresCounter: attestationFailuresCounter,
+		kdsGetter:                  kdsGetter,
+	}, ticker.Stop
+}
+
+// ServerHandshake implements an aTLS handshake for the latest state.
+//
+// If successful, the state will be passed to gRPC as AuthInfo.
```

Comment:
```suggestion
// If successful, the state will be passed to gRPC as [AuthInfo].
```
nit

Reply:
Are these brackets actually interpreted by something, or do you suggest them for readability?

Reply:
They are interpreted by `gopls`, and, when hovering, [show the context information for that type](https://tip.golang.org/doc/comment#package).

Reply:
Thanks, adapted.